ui/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/DeploymentService.scala

```diff
@@ -58,22 +64,44 @@ class DeploymentService(processRepository: FetchingProcessRepository[Future],
   }
 
   //inner Future in result allows to wait for deployment finish, while outer handles validation
-  def deployProcess(processId: ProcessId,
+  def deployProcess(processIdWithName: ProcessIdWithName,
                     savepointPath: Option[String],
                     deploymentComment: Option[DeploymentComment],
                     deploymentManager: ProcessingType => DeploymentManager)
                    (implicit user: LoggedUser): Future[Future[ProcessActionEntityData]] = {
-    for {
-      maybeProcess <- processRepository.fetchLatestProcessDetailsForProcessId[CanonicalProcess](processId)
-      process <- processDataExistOrFail(maybeProcess, processId.value.toString)
-      result <- deployAndSaveProcess(process, savepointPath, deploymentComment, deploymentManager(process.processingType))
-    } yield result
+    val result = for {
+        maybeProcess <- processRepository.fetchLatestProcessDetailsForProcessId[CanonicalProcess](processIdWithName.id)
+        process <- processDataExistOrFail(maybeProcess, processIdWithName.id)
+        result <- deployAndSaveProcess(process, savepointPath, deploymentComment, deploymentManager(process.processingType))
+      } yield result
+    notifyOfDeploymentAction(processIdWithName, "deploy", deploymentComment)(result.flatten)
+    result
   }
 
-  private def processDataExistOrFail[T](maybeProcess: Option[T], processId: String): Future[T] = {
+  //TODO: there is small problem here: if no one invokes process status for long time, Flink can remove process from history
+  //- then it's gone, not finished.
+  def handleFinishedProcess(idWithName: ProcessIdWithName, processState: Option[ProcessState]): Future[Unit] = {
+    implicit val user: NussknackerInternalUser.type = NussknackerInternalUser
+    implicit val listenerUser: ListenerUser = ListenerApiUser(user)
+    processState match {
+      case Some(state) if state.status.isFinished =>
+        findDeployedVersion(idWithName).flatMap {
+          case Some(version) => {
+            val finishedDeploymentComment = DeploymentComment.unsafe("Scenario finished")
+            actionRepository.markProcessAsCancelled(idWithName.id, version, Some(finishedDeploymentComment)).map(_ =>
+              processChangeListener.handle(OnFinished(idWithName.id, version))
+            )
+          }
+          case _ => Future.successful(())
+        }
+      case _ => Future.successful(())
+    }
+  }
+
+  private def processDataExistOrFail[T](maybeProcess: Option[T], processId: ProcessId): Future[T] = {
     maybeProcess match {
       case Some(processData) => Future.successful(processData)
-      case None => Future.failed(ProcessNotFoundError(processId))
+      case None => Future.failed(ProcessNotFoundError(processId.value.toString))
```

Comment:
Why `ProcessNotFoundError` accepts string, not VO?

Reply:
I'll fix it in next PR...